reactioncommerce/reaction
# #4353 Plugin for auto-generated sitemaps

Resolves #4353
Impact: **minor**  
Type: **feature**

## Issue
Reaction currently has no sitemap.xml - a file that allows search engines visibility into a website's pages. Without it, search engines must manually crawl and find pages through links.

## Solution
A sitemap generator plugin that creates and stores XML for a sitemap index, as well as sitemaps for tag pages, PDPs, and arbitrary URLs that can be added via an event hook.

There is a recurring job that runs every 24 hours (can be changed) that generates the sitemaps. There is also a button to manually trigger a refresh, at Dashboard -> Shop -> Options - along with a notification that appears when it's completed.

GraphQL TBD (separate issue)

## Breaking changes
None

## Testing
1. Reset the db (translations updated) and use `reaction-data-loader` to load the dev dataset
2. Create some products and tags - some archived/non-visible, some visible/published
3. Login as admin and visit Dashboard -> Shop Settings -> Options
4. Click "Refresh sitemap"
5. Confirm you get a notification (bell icon) as well as a Toast alert.
6. See Sitemaps collection and confirm correct XML by handle: sitemap.xml, sitemap-pages-1.xml, sitemap-tags-1.xml, sitemap-products-1.xml
7. Visit /sitemap.xml and confirm it loads and displays sitemap index XML
8. Visit each of the individual sitemap URLs and confirm they load/display valid XML
9. In the DB, confirm you see a Sitemaps collection with a document for each sitemap.
10. Confirm you see "BASE_URL" in place of any actual URLs in DB sitemaps
11. Without updating a tag or product, click the sitemap refresh button again and confirm that the sitemaps for tags and products did not regenerate, but that the sitemap index (sitemap.xml) and the one for pages, has regenerated.
12. Update a tag's title (I did it by editing the main tag nav), refresh sitemap, and confirm the tag sitemaps did regenerate, but not product sitemaps
13. Update a product field (such as the title), refresh sitemap, and confirm product sitemaps did regenerate
14. Run against retailer and miderprise databases - confirm job runs in background (doesn't stop app UI from functioning), and creates the correct number of sitemaps (1k URLs per sitemap max).
15. At Shop -> Options, change the "Sitemap refresh period" value to something really short, like "every 30 seconds"
16. Restart the server for the job schedule to update (this appears to be the same as other refresh period settings, like open exchange rates)
17. Confirm the job runs as scheduled.
